src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java

```diff
@@ -33,45 +33,88 @@
 public class CISShadowMonitor implements Consumer<NetworkStateProvider.ConnectionState> {
     private static final Logger LOGGER = LogManager.getLogger(CISShadowMonitor.class);
     private static final String CIS_SHADOW_SUFFIX = "-gci";
     private static final String VERSION = "version";
-    private static final long TIMEOUT_FOR_SUBSCRIBING_TO_TOPICS_SECONDS = Duration.ofMinutes(1).getSeconds();
-    private static final long WAIT_TIME_TO_SUBSCRIBE_AGAIN_IN_MS = Duration.ofMinutes(2).toMillis();
-    private static final Random JITTER = new Random();
-    static final String SHADOW_UPDATE_DELTA_TOPIC = "$aws/things/%s/shadow/update/delta";
-    static final String SHADOW_GET_ACCEPTED_TOPIC = "$aws/things/%s/shadow/get/accepted";
 
+    private static final RetryUtils.RetryConfig SUBSCRIBE_TO_TOPICS_RETRY_CONFIG =
+            RetryUtils.RetryConfig.builder()
+                    .initialRetryInterval(Duration.ofSeconds(1L))
+                    .maxRetryInterval(Duration.ofSeconds(30L))
+                    .maxAttempt(Integer.MAX_VALUE)
+                    .retryableExceptions(Collections.singletonList(Exception.class))
+                    .build();
     private static final RetryUtils.RetryConfig GET_CONNECTIVITY_RETRY_CONFIG =
             RetryUtils.RetryConfig.builder().initialRetryInterval(Duration.ofMinutes(1L))
                     .maxRetryInterval(Duration.ofMinutes(30L)).maxAttempt(Integer.MAX_VALUE)
                     .retryableExceptions(Arrays.asList(ThrottlingException.class, InternalServerException.class))
                     .build();
+    private static final RetryUtils.RetryConfig GET_CIS_SHADOW_RETRY_CONFIG =
+            RetryUtils.RetryConfig.builder()
+                    .initialRetryInterval(Duration.ofSeconds(1L))
+                    .maxRetryInterval(Duration.ofSeconds(30L))
+                    .maxAttempt(Integer.MAX_VALUE)
+                    .retryableExceptions(Collections.singletonList(Exception.class))
+                    .build();
+
+    // we only need to subscribe to shadow topics once across the lifetime of this monitor
+    private final AtomicBoolean subscribed = new AtomicBoolean();
+
+    // we don't need to unsubscribe, because mqtt client reconnects with clean session on startup.
+    // instead, we can cancel current operations and block new ones on shutdown.
+    private final AtomicBoolean stopped = new AtomicBoolean();
+    private final Consumer<ShadowDeltaUpdatedEvent> onShadowDeltaUpdated = event -> {
+        if (stopped.get()) {
+            return;
+        }
+        processCISShadow(event);
+    };
+    private final Consumer<GetShadowResponse> onGetShadowAccepted = resp -> {
+        if (stopped.get()) {
+            return;
+        }
+        reportShadowReceived();
+        processCISShadow(resp);
+    };
+    private final Consumer<ErrorResponse> onGetShadowRejected = err -> {
+        if (stopped.get()) {
+            return;
+        }
+        reportShadowReceived();
```

Comment:
report shadow received? But this is rejected...

Reply:
bad naming, it's to signal that we've received any shadow response, whether accepted or rejected, will think of better name